wcfsetup/install/files/lib/http/error/NotFoundHandler.class.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace wcf\http\error;
+
+use Laminas\Diactoros\Response\HtmlResponse;
+use Laminas\Diactoros\Response\JsonResponse;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use wcf\http\Helper;
+use wcf\system\box\BoxHandler;
+use wcf\system\session\SessionHandler;
+use wcf\system\WCF;
+
+/**
+ * Returns a "Not Found" response.
+ *
+ * @author  Tim Duesterhus
+ * @copyright   2001-2023 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @since   6.0
+ */
+final class NotFoundHandler implements RequestHandlerInterface
+{
+    public function handle(ServerRequestInterface $request): ResponseInterface
+    {
+        $errorDetail = ErrorDetail::fromRequest($request);
+        $message = $errorDetail?->getMessage() ?? WCF::getLanguage()->getDynamicVariable('wcf.page.error.illegalLink');
+
+        if (!\class_exists(WCFACP::class, false)) {
+            BoxHandler::disablePageLayout();
+        }
+        SessionHandler::getInstance()->disableTracking();
+
+        $preferredType = Helper::getPreferredContentType($request, [
+            'text/html',
+            'application/json',
```

Comment:
These should be swapped.
* A regular browser request will ask for HTML.
* AJAX requests will expect JSON (at least in case of our error handling).
* AJAX requests could ask for a different content type, but in case of an error we’re returning JSON.

For all other types of requests, where HTML was not requested, HTML is not a valid substitute either. For example, an image tag will refuse to load anything that is not an image, therefore we can return whatever we want, because it will not be a valid substitute either way.

Reply:
I believe that an HTML response is the correct default if the client does not send a preference. I've added a FixAcceptHeader middleware instead that adds a preference for JSON for AJAX requests without any preference. What do you think?

Reply:
You’re missing my point here: Unless the client explicitly (*) requests HTML, HTML is _never_ a valid substitute. You are then either dealing with requests for other media types, for example, images or videos, in which case JSON at least saves computing time and bandwidth on the server, or any other type of request that is most likely a programmatic request.

(*) Or implicitly through the proposed `FixAcceptHeader` middleware.